File: components/heishamon/select.cpp
#include "esphome/core/defines.h"
#ifdef USE_SELECT
#include "select.h"
#include "esphome/core/log.h"
#include <map>

namespace esphome {
namespace heishamon {

static const char *const TAG = "heishamon.select";

// Maps for converting option strings to numeric command values
static const std::map<std::string, std::map<std::string, uint8_t>> SELECT_VALUE_MAPS = {
    // SET3 - Quiet Mode
    {"quiet_mode", {
        {"Off", 0}, {"Level 1", 1}, {"Level 2", 2}, {"Level 3", 3}, {"Scheduled", 4}
    }},
    // SET4 - Powerful Mode (0/30/60/90 min)
    {"powerful_mode", {
        {"Off", 0}, {"30 min", 1}, {"60 min", 2}, {"90 min", 3}
    }},
    // SET9 - Operating Mode
    {"operating_mode", {
        {"Heat only", 0}, {"Cool only", 1}, {"Auto(Heat)", 2}, {"DHW only", 3},
        {"Heat+DHW", 4}, {"Cool+DHW", 5}, {"Auto(Heat)+DHW", 6},
        {"Auto(Cool)", 7}, {"Auto(Cool)+DHW", 8}
    }},
    // SET17 - Active Zones
    {"zones", {
        {"Zone 1 only", 0}, {"Zone 2 only", 1}, {"Zone 1 + Zone 2", 2}
    }},
    // SET35 - Bivalent Mode
    {"bivalent_mode", {
        {"Alternative", 0}, {"Parallel", 1}, {"Advanced Parallel", 2}
    }},
    // SET26 - External Pad Heater
    {"external_pad_heater", {
        {"Disabled", 0}, {"Type-A", 1}, {"Type-B", 2}
    }},
    // Smart Grid Mode
    {"smart_grid_mode", {
        {"Normal", 0}, {"Capacity 1", 1}, {"HP/DHW Off", 2}, {"Capacity 2", 3}
    }},
    // Read-only selects (for decoding received values)
    {"heating_mode", {
        {"Compensation curve", 0}, {"Direct", 1}
    }},
    {"cooling_mode", {
        {"Compensation curve", 0}, {"Direct", 1}
    }},
    {"three_way_valve", {
        {"Room", 0}, {"DHW", 1}
    }},
    {"holiday_mode_state", {
        {"Off", 0}, {"Scheduled", 1}, {"Active", 2}
    }},
    {"solar_mode", {
        {"Disabled", 0}, {"Buffer", 1}, {"DHW", 2}
    }},
    {"pump_flowrate_mode", {
        {"DeltaT", 0}, {"Max flow", 1}
    }},
    {"liquid_type", {
        {"Water", 0}, {"Glycol", 1}
    }},
    {"z1_sensor_settings", {
        {"Water temperature", 0}, {"External thermostat", 1}, {"Internal thermostat", 2}, {"Thermistor", 3}
    }},
    {"z2_sensor_settings", {
        {"Water temperature", 0}, {"External thermostat", 1}, {"Internal thermostat", 2}, {"Thermistor", 3}
    }},
    // SET43 - DHW Sensor Selection (K/L series All-In-One)
    {"dhw_sensor_selection", {
        {"Top", 0}, {"Center", 1}
    }},
    // SET44 - DHW Heater State
    {"dhw_heater_state", {
        {"Blocked", 0}, {"Free", 1}
    }},
    // SET45 - Room Heater State
    {"room_heater_state", {
        {"Blocked", 0}, {"Free", 1}
    }},
};

// Reverse map for converting numeric values to option strings
static std::string value_to_option(const std::string &select_type, uint8_t value) {
    auto type_it = SELECT_VALUE_MAPS.find(select_type);
    if (type_it == SELECT_VALUE_MAPS.end()) {
        return "";
    }
    for (const auto &pair : type_it->second) {
        if (pair.second == value) {
            return pair.first;
        }
    }
    return "";
}

void HeishamonSelect::setup() {
  ESP_LOGCONFIG(TAG, "Setting up Heishamon Select '%s'...", this->select_type_.c_str());
  
  // Register callback for topic-based updates (works in both modes)
  if (!this->topic_.empty() && this->parent_ != nullptr) {
    ESP_LOGCONFIG(TAG, "Registering callback for topic: %s", this->topic_.c_str());
    this->parent_->register_select_callback(this->topic_, [this](const std::string &value) {
      ESP_LOGD(TAG, "Received value for %s: %s", this->select_type_.c_str(), value.c_str());
      // Try to convert numeric value to option string
      char *end;
      long num_value = strtol(value.c_str(), &end, 10);
      if (*end == '\0') {
        // It's a numeric value, convert to option string
        std::string option = value_to_option(this->select_type_, static_cast<uint8_t>(num_value));
        if (!option.empty()) {
          this->publish_state(option);
          return;
        }
      }
      // Otherwise use value as-is (already an option string)
      this->publish_state(value);
    });
  }
  
  // Set initial option if specified
  if (!this->initial_option_.empty()) {
    this->publish_state(this->initial_option_);
  }
}

void HeishamonSelect::dump_config() {
  ESP_LOGCONFIG(TAG, "Heishamon Select:");
  ESP_LOGCONFIG(TAG, "  Type: %s", this->select_type_.c_str());
  if (this->listen_only_ && !this->topic_.empty()) {
    ESP_LOGCONFIG(TAG, "  Mode: Listen-only");
    ESP_LOGCONFIG(TAG, "  Topic: %s", this->topic_.c_str());
  } else {
    ESP_LOGCONFIG(TAG, "  Mode: Command-based");
    ESP_LOGCONFIG(TAG, "  Command: %s", this->command_.c_str());
  }
}

void HeishamonSelect::control(const std::string &value) {
  ESP_LOGD(TAG, "HeishamonSelect control called with value: %s for type: %s", 
           value.c_str(), this->select_type_.c_str());
  
  // In listen-only mode, just log and don't send command
  if (this->listen_only_) {
    ESP_LOGW(TAG, "Select '%s' is read-only in listen-only mode", this->select_type_.c_str());
    return;
  }
  
  // Convert option string to numeric value
  auto type_it = SELECT_VALUE_MAPS.find(this->select_type_);
  if (type_it == SELECT_VALUE_MAPS.end()) {
    ESP_LOGE(TAG, "Unknown select type: %s", this->select_type_.c_str());
    return;
  }
  
  auto value_it = type_it->second.find(value);
  if (value_it == type_it->second.end()) {
    ESP_LOGE(TAG, "Unknown option '%s' for select type '%s'", value.c_str(), this->select_type_.c_str());
    return;
  }
  
  uint8_t cmd_value = value_it->second;
  ESP_LOGD(TAG, "Sending command %s with value %d", this->command_.c_str(), cmd_value);
  
  // Send command via parent component
  if (this->parent_ != nullptr) {
    this->parent_->send_command(this->command_, cmd_value);
    // Optimistically update state
    this->publish_state(value);
  } else {
    ESP_LOGE(TAG, "Parent component not set, cannot send command");
  }
}

}  // namespace heishamon
}  // namespace esphome
#endif

File: components/heishamon/select.py
import esphome.codegen as cg
import esphome.config_validation as cv
from esphome.components import select
from esphome.const import CONF_ID
from . import HeishamonComponent, heishamon_ns

DEPENDENCIES = ["heishamon", "select"]
CODEOWNERS = ["@AntorFR"]

HeishamonSelect = heishamon_ns.class_("HeishamonSelect", select.Select, cg.Component)

# Available select entities based on HeishaMon protocol analysis
# Reference: https://github.com/Egyras/HeishaMon and heishamon-homeassistant
HEISHA_SELECT_OPTIONS = {
    # ============ CONTROL SELECTS (can send commands) ============
    
    # SET3 - Quiet Mode (0-4, where 4=scheduled)
    "quiet_mode": {
        "name": "Quiet Mode",
        "command": "SetQuietMode",
        "topic": "quiet_mode_level",  # Also listen to current state
        "options": ["Off", "Level 1", "Level 2", "Level 3", "Scheduled"],
        "default": "Off",
        "icon": "mdi:volume-off"
    },
    
    # SET4 - Powerful Mode (0-3 = 0/30/60/90 minutes)
    "powerful_mode": {
        "name": "Powerful Mode",
        "command": "SetPowerfulMode",
        "topic": "powerful_mode_time",
        "options": ["Off", "30 min", "60 min", "90 min"],
        "default": "Off",
        "icon": "mdi:rocket-launch"
    },
    
    # SET9 - Operating Mode (0-8)
    "operating_mode": {
        "name": "Operating Mode",
        "command": "SetOperationMode",
        "topic": "operation_mode",
        "options": [
            "Heat only",        # 0
            "Cool only",        # 1
            "Auto(Heat)",       # 2
            "DHW only",         # 3
            "Heat+DHW",         # 4
            "Cool+DHW",         # 5
            "Auto(Heat)+DHW",   # 6
            "Auto(Cool)",       # 7
            "Auto(Cool)+DHW"    # 8
        ],
        "default": "Heat+DHW",
        "icon": "mdi:heat-pump"
    },
    
    # SET17 - Active Zones (0-2)
    "zones": {
        "name": "Active Zones",
        "command": "SetZones",
        "topic": "zones_state",
        "options": ["Zone 1 only", "Zone 2 only", "Zone 1 + Zone 2"],
        "default": "Zone 1 only",
        "icon": "mdi:home-thermometer"
    },
    
    # SET35 - Bivalent Mode (0-2)
    "bivalent_mode": {
        "name": "Bivalent Mode",
        "command": "SetBivalentMode",
        "topic": "bivalent_mode",
        "options": ["Alternative", "Parallel", "Advanced Parallel"],
        "default": "Alternative",
        "icon": "mdi:heat-wave"
    },
    
    # SET26 - External Pad Heater Type (0-2)
    "external_pad_heater": {
        "name": "External Pad Heater",
        "command": "SetExternalPadHeater",
        "topic": "external_pad_heater",
        "options": ["Disabled", "Type-A", "Type-B"],
        "default": "Disabled",
        "icon": "mdi:radiator"
    },
    
    # Smart Grid Mode (Optional PCB command)
    "smart_grid_mode": {
        "name": "Smart Grid Mode",
        "command": "SetSmartGridMode", 
        "topic": "smart_grid_mode",
        "options": ["Normal", "Capacity 1", "HP/DHW Off", "Capacity 2"],
        "default": "Normal",
        "icon": "mdi:transmission-tower"
    },
    
    # ============ READ-ONLY SELECTS (topic-based, no command) ============
    
    # Heating Mode - Read from TOP76
    "heating_mode": {
        "name": "Heating Mode",
        "topic": "heating_mode",
        "options": ["Compensation curve", "Direct"],
        "default": "Compensation curve",
        "icon": "mdi:thermometer-lines"
    },
    
    # Cooling Mode - Read from TOP81
    "cooling_mode": {
        "name": "Cooling Mode", 
        "topic": "cooling_mode",
        "options": ["Compensation curve", "Direct"],
        "default": "Compensation curve",
        "icon": "mdi:snowflake-thermometer"
    },
    
    # 3-Way Valve State - Read from TOP20
    "three_way_valve": {
        "name": "3-Way Valve State",
        "topic": "three_way_valve_state",
        "options": ["Room", "DHW"],
        "default": "Room",
        "icon": "mdi:valve"
    },
    
    # Holiday Mode State - Read from TOP19
    "holiday_mode_state": {
        "name": "Holiday Mode State",
        "topic": "holiday_mode_state",
        "options": ["Off", "Scheduled", "Active"],
        "default": "Off",
        "icon": "mdi:beach"
    },
    
    # Solar Mode - Read from TOP101
    "solar_mode": {
        "name": "Solar Mode",
        "topic": "solar_mode",
        "options": ["Disabled", "Buffer", "DHW"],
        "default": "Disabled",
        "icon": "mdi:solar-power"
    },
    
    # Pump Flowrate Mode - Read from TOP106 (J-series)
    "pump_flowrate_mode": {
        "name": "Pump Flowrate Mode",
        "topic": "pump_flowrate_mode",
        "options": ["DeltaT", "Max flow"],
        "default": "DeltaT",
        "icon": "mdi:pump"
    },
    
    # Liquid Type - Read from TOP107
    "liquid_type": {
        "name": "Liquid Type",
        "topic": "liquid_type",
        "options": ["Water", "Glycol"],
        "default": "Water",
        "icon": "mdi:water"
    },
    
    # Zone 1 Sensor Settings - Read from TOP112
    "z1_sensor_settings": {
        "name": "Zone 1 Sensor",
        "topic": "z1_sensor_settings",
        "options": ["Water temperature", "External thermostat", "Internal thermostat", "Thermistor"],
        "default": "Water temperature",
        "icon": "mdi:thermometer"
    },
    
    # Zone 2 Sensor Settings - Read from TOP111
    "z2_sensor_settings": {
        "name": "Zone 2 Sensor",
        "topic": "z2_sensor_settings",
        "options": ["Water temperature", "External thermostat", "Internal thermostat", "Thermistor"],
        "default": "Water temperature",
        "icon": "mdi:thermometer"
    },
}

CONFIG_SCHEMA = cv.Schema(
    {
        cv.GenerateID(): cv.declare_id(HeishamonSelect),
        cv.GenerateID("heishamon_id"): cv.use_id(HeishamonComponent),
        cv.Required("type"): cv.one_of(*HEISHA_SELECT_OPTIONS.keys(), lower=True),
    }
).extend(select.select_schema(HeishamonSelect))

async def to_code(config):
    var = cg.new_Pvariable(config[CONF_ID])
    await cg.register_component(var, config)
    
    # Get options for this select type
    select_config = HEISHA_SELECT_OPTIONS[config["type"]]
    options = select_config["options"]
    
    # register_select handles options via traits
    await select.register_select(var, config, options=options)

    parent = await cg.get_variable(config["heishamon_id"])
    cg.add(var.set_parent(parent))
    cg.add(var.set_select_type(config["type"]))
    
    # Configure mode based on select type
    cg.add(var.set_initial_option(select_config["default"]))
    
    # Check if this has a command (control mode) or just topic (read-only)
    if "command" in select_config:
        # Command-based select for control mode
        cg.add(var.set_command(select_config["command"]))
        cg.add(var.set_listen_only(False))
        # Also set topic if available for reading current state
        if "topic" in select_config:
            cg.add(var.set_topic(select_config["topic"]))
    else:
        # Topic-based select for listen-only mode (no command)
        cg.add(var.set_topic(select_config["topic"]))
        cg.add(var.set_listen_only(True))
